### PR TITLE
Use containers/image lib instead of docker 

### LINKED ISSUE #35: Use containers/image library for docker-less pull and mount
Requiring access to the docker.socket is a lot of privilege. With the https://github.com/containers/image/ library this should not be required.

### PR DESCRIPTION
**Pull request by ilackarms**
This will use github.com/containers/image to pull images instead of connecting to the docker socket (which requires running  in privileged mode).

Use of the docker daemon is still possible via `--use-docker`

Fixes #35

(this PR is an update of #37)

### CODE REVIEW

**Comment by simon3z** [comment]
> Use of the docker daemon is still possible via --use-docker

@ilackarms I am not sure if we should break the cli backward-compatibility. We should probably use `--no-docker` (or similar) instead.

**Comment by ilackarms** [comment]
@mtrmac can you please review  https://github.com/openshift/image-inspector/pull/58/commits/50efed132182ea4177096ccc4042fb7e8f08fa13 to see that i have correctly understood whiteout file format in layers?

**Comment by simon3z** [comment]
@ilackarms do we really need the go 1.8 dependency?

**Comment by ilackarms** [comment]
@simon3z sorry that was left in from a debug build; removing it now

**Comment by mtrmac** [comment]
> @mtrmac can you please review 50efed1 to see that i have correctly understood whiteout file format in layers?

All I know about whiteouts is that they exist; perhaps @runcom can help or point to someone knowledgeable?

File: pkg/inspector/image-inspector.go [hunk 1]
```diff
@@ -41,6 +43,8 @@ const (
 	CHROOT_SERVE_PATH        = "/"
 	OSCAP_CVE_DIR            = "/tmp"
 	PULL_LOG_INTERVAL_SEC    = 10
+	DOCKER_CERTS_DIR         = "/etc/docker/certs.d"
+	DEFAULT_SIGN_POLICY      = "{\"default\":[{\"type\": \"insecureAcceptAnything\" }]}"
```
Comment (mtrmac) [inline pkg/inspector/image-inspector.go]: Nit, this is now unused.

File: pkg/inspector/image-inspector.go [hunk 2]
```diff
@@ -540,3 +486,186 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+
+	// Try all the possible auth's from the config file
+	var authErr error
+	log.Println("Copying image")
+	for name, auth := range imagePullAuths.Configs {
+		sourceCtx.DockerAuthConfig = &types.DockerAuthConfig{
+			Username: auth.Username,
+			Password: auth.Password,
+		}
+		authErr = copy.Image(policyContext, destRef, srcRef, &copy.Options{
```
Comment (mtrmac) [inline pkg/inspector/image-inspector.go]: BTW, because the code only streams each layer once, a `copy.Image` and a full on-disk temporary copy may not be necessary: Collect layer infos using `img.LayerInfos()`, and then use `ImageSource.GetBlob()` to get `io.Reader`s for each of the tar layers.

OTOH using `copy.Image` gives you (signature verification if enabled), downloaded blob verification against the expected digests, and progress reporting for free, and the temporary on-disk image is easier to inspect when debugging.

Either works, depends on what your project needs more.

**Review by mtrmac — COMMENTED** [review]
Two more minor notes.

https://github.com/openshift/image-inspector/pull/37/files/f9b60d813e0e2ec1457a13153e77c010bbadae47#r122797178 may still be worth considering.

File: pkg/inspector/image-inspector.go (hunk 2 above)
Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: I think signature verification is a nice feature to have; I wouldn't be opposed to writing the layer twice to disk (compressed and then uncompressed) for the benefit of having signature inspection. thoughts, @simon3z ?

Comment (mtrmac) [inline pkg/inspector/image-inspector.go]: (Signature verification can be done without `copy.Image` as well; but of course the more features of `copy.Image` are required, the more attractive it is to call the single higher-level function.)

**Comment by simon3z** [comment]
@ilackarms maybe it's time to squash commits (I am not saying to squash them all in one... but at least few).

Also, did you run `dep prune` I am worried about those 7000 files :blush:

**Comment by ilackarms** [comment]
@simon3z I can squash now, although I usually look to get the *lgtm* before squashing, in case i need to revert something

**Comment by simon3z** [comment]
OK let's go in order... I see that this is including (of course) #57 and #38. Let's get those merged first so this will look cleaner I suppose.

**Comment by ilackarms** [comment]
@simon3z @enoodle @pweil- please (re-)review

**Comment by pweil-** [comment]
@mfojtik PTAL

**Comment by mfojtik** [comment]
@ilackarms @simon3z i would keep the dep bumps as separate commits next time, it makes it easier to review :-)

File: Dockerfile [hunk 3]
```diff
@@ -2,17 +2,28 @@ FROM centos:7
...
-    yum install -y golang openscap-scanner && \
+    yum install -y \
+    openscap-scanner && \
     yum clean all
 
+    #needs go 1.7 for context
+RUN cd /tmp && \
+    curl -LO https://storage.googleapis.com/golang/go1.7.linux-amd64.tar.gz && \
+    tar -C /usr/local -xvzf go1.7.linux-amd64.tar.gz && \
+    rm -f go1.7.linux-amd64.tar.gz
```
Comment (mfojtik) [inline Dockerfile]: nit: can we squash the lines below to this and remove the go? i guess it adds ~100MB to the image this way :-)

File: pkg/inspector/image-inspector.go [hunk 4]
```diff
@@ -103,25 +104,33 @@ func NewDefaultImageInspector(opts iicmd.ImageInspectorOptions) ImageInspector {
...
 // Inspect inspects and serves the image based on the ImageInspectorOptions.
 func (i *defaultImageInspector) Inspect() error {
-	client, err := docker.NewClient(i.opts.URI)
-	if err != nil {
-		return fmt.Errorf("Unable to connect to docker daemon: %v\n", err)
-	}
+	if i.opts.UseDockerSocket {
+		client, err := docker.NewClient(i.opts.DockerSocket)
+		if err != nil {
+			return fmt.Errorf("Unable to connect to docker daemon: %v\n", err)
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: nit: in origin we have convention that fmt.Errorf is always lowecase as you then prefix them with 'ERROR:' or something at higher level.

File: pkg/inspector/image-inspector.go [hunk 5]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+			return nil, "", fmt.Errorf("Invalid source name %s: %v", i.opts.Image, err)
+		}
+	}
+
+	certPath, err := i.certPath(i.opts.Image)
+	if err != nil {
+		return nil, "", err
+	}
+	sourceCtx := &types.SystemContext{
+		DockerAuthConfig: nil,
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: no need to set nil field?

File: pkg/inspector/image-inspector.go [hunk 6]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+	}
+
+	imagePullAuths, err := i.getAuthConfigs()
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Try all the possible auth's from the config file
+	var authErr error
+	log.Println("Copying image")
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: i wish we can controller the verbosity at some point... for now the convention we have for progress messages is to append " ..." at the end ("Copying image ...")

File: pkg/inspector/image-inspector.go [hunk 7]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+		return nil, "", fmt.Errorf("Error while extracting downloaded image: %v\n", err)
+	}
+
+	return inspectInfo, imageDigest, nil
+}
+
+// extractDownloadedImage will untar all the layer tar files specified in 'info'
+func (i *defaultImageInspector) extractDownloadedImage(layers []string) error {
+	for _, layer := range layers {
+		filename := i.opts.DstPath + "/" + strings.SplitN(layer, ":", 2)[1] + ".tar"
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: filepath.Join()

File: pkg/inspector/image-inspector.go [hunk 8]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+		log.Printf("Untar %s\n", filename)
+		if err := util.UntarGzLayer(filename, destDir); err != nil {
+			return err
+		}
+	}
+	log.Printf("finished extraction")
+	return nil
+}
+
+// dockerCertPath will try to extract the registry name from the image and return
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: godoc

File: pkg/inspector/image-inspector.go [hunk 9]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+// "/etc/docker/certs.d/<RERGISTRY_NAME>" if this path exists or nil otherwise.
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: typo

File: pkg/inspector/image-inspector.go [hunk 10]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+func (i *defaultImageInspector) certPath(fullImageName string) (string, error) {
+	if len(i.opts.RegistryCertPath) > 0 {
+		return i.opts.RegistryCertPath, nil
+	}
+
+	// try to find certificates from docker
+	source_name := strings.SplitN(fullImageName, "://", 2)
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: check the len (or we panic when the image name is malformed)

File: pkg/inspector/image-inspector.go [hunk 11]
```diff
@@ -541,3 +485,154 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+	name := source_name[len(source_name)-1]
+	names := strings.SplitN(name, "/", 2)
+	certsPath := DOCKER_CERTS_DIR + "/" + names[0]
+	_, err := os.Stat(certsPath)
+	if err != nil {
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: if _, err := os.Stat(certsPath); os.IsNotExist(err) { ... }

File: pkg/inspector/image-inspector_test.go [hunk 12]
```diff
@@ -5,11 +5,11 @@ import (
 	"io"
 	"io/ioutil"
 	"os"
-	"testing"
 
 	docker "github.com/fsouza/go-dockerclient"
 	iiapi "github.com/openshift/image-inspector/pkg/api"
 	iicmd "github.com/openshift/image-inspector/pkg/cmd"
+	"testing"
```
Comment (mfojtik) [inline pkg/inspector/image-inspector_test.go]: why this change?

File: pkg/util/util.go [hunk 13]
```diff
@@ -1,5 +1,21 @@
...
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	DOCKER_TAR_PREFIX = "rootfs/"
```
Comment (mfojtik) [inline pkg/util/util.go]: nit: In general, we use capitalized constants in Go :-)

File: cmd/image-inspector.go [hunk 14]
```diff
@@ -15,7 +15,8 @@ import (
 func main() {
 	inspectorOptions := iicmd.NewDefaultImageInspectorOptions()
 
-	flag.StringVar(&inspectorOptions.URI, "docker", inspectorOptions.URI, "Daemon socket to connect to")
+	flag.BoolVar(&inspectorOptions.UseDockerSocket, "use-docker-socket", inspectorOptions.UseDockerSocket, "Use docker daemon to handle image pulls (requires docker-socket)")
```
Comment (pweil-) [inline cmd/image-inspector.go]: do we need this parameter or can we imply that if `docker-socket` is set then we are using it and if not we aren't?

File: pkg/util/util.go [hunk 15]
```diff
@@ -13,3 +29,107 @@ func Min(x, y int) int {
...
+
+		hdrInfo := hdr.FileInfo()
+
+		dstpath := path.Join(destination, strings.TrimPrefix(hdr.Name, DOCKER_TAR_PREFIX))
+		// Overriding permissions to allow writing content
+		mode := hdrInfo.Mode() | OWNER_PERM_RW
+
+		// opaque whiteout file
+		// https://github.com/opencontainers/image-spec/blob/master/layer.md#opaque-whiteout
+		if strings.HasSuffix(dstpath, ".wh..wh..opq") {
```
Comment (mfojtik) [inline pkg/util/util.go]: i wish they have constant for this suffix.

File: pkg/util/util.go [hunk 16]
```diff
@@ -13,3 +29,107 @@ func Min(x, y int) int {
...
+			// For now we're skipping anything else. Special device files and
+			// symlinks are not needed or anyway probably incorrect.
+		}
+
+		// maintaining access and modification time in best effort fashion
+		os.Chtimes(dstpath, hdr.AccessTime, hdr.ModTime)
+	}
+}
+
+func UntarGzLayer(filename, destination string) error {
```
Comment (mfojtik) [inline pkg/util/util.go]: add godoc

**Comment by mfojtik** [comment]
some nits but LGTM

File: pkg/inspector/image-inspector_test.go (hunk 12 above)
Comment (pweil-) [inline pkg/inspector/image-inspector_test.go]: change is no-op

File: pkg/util/util.go (hunk 13 above)
Comment (pweil-) [inline pkg/util/util.go]: godoc please

File: Gopkg.lock [hunk 17]
```diff
@@ -2,14 +2,175 @@
 
 
 [[projects]]
+  branch = "master"
+  name = "github.com/Azure/go-ansiterm"
+  packages = [".","winterm"]
+  revision = "fa152c58bc15761d0200cb75fe958b89a9d4888e"
+
+[[projects]]
+  name = "github.com/Microsoft/go-winio"
```
Comment (pweil-) [inline Gopkg.lock]: Seeing these makes me wonder if we need to have something split into a separate interface with build-tag include implementations.  What are these dependencies being used for?

File: cmd/image-inspector.go (hunk 14 above)
Comment (ilackarms) [inline cmd/image-inspector.go]: `docker-socket` is set by default (to `"unix:///var/run/docker.sock"`)

File: Dockerfile (hunk 3 above)
Comment (ilackarms) [inline Dockerfile]: we need go 1.7 which is not currently available as an rpm

File: cmd/image-inspector.go (hunk 14 above)
Comment (pweil-) [inline cmd/image-inspector.go]: hmm, yeah...

File: Dockerfile (hunk 3 above)
Comment (mfojtik) [inline Dockerfile]: that is fine, but I don't think you need it for the runtime right? just to build the image-inspector and after that point it is just a garbage in the image ;)

File: Gopkg.lock (hunk 17 above)
Comment (ilackarms) [inline Gopkg.lock]: we use [golang/dep](https://github.com/golang/dep) to vendor dependencies currently. even though `go-winio` is only used when `GOOS=windows` or using `-tags=windows` which I don't think is a supported use case, the vendoring tool doesn't know that and imports any secondary dependencies (this one is from https://github.com/ilackarms/image-inspector/blob/dbae546537dcfcc4d9b55d66fb2646a1add513ea/vendor/github.com/containers/storage/pkg/system/filesys_windows.go)

File: pkg/util/util.go (hunk 13 above)
Comment (ilackarms) [inline pkg/util/util.go]: should be private instead

File: Gopkg.lock (hunk 17 above)
Comment (mtrmac) [inline Gopkg.lock]: FWIW this is ultimately dragged in by using `alltransports.ParseImageName`, which roughly amounts to “I want to be able to consume any kind of image, let the user specify a `"`_transport_`:`_something_`"` string which I don’t understand and just pass along to the library and let the library do everything that is needed”.

I do think that such an ability to inspect an image wherever it is stored can be very useful, but if you were definitely restricted to docker/distribution-hosted images by some other factor, replace `alltransports.ParseImageName` with `containers/image/docker.ParseReference` (using the after-`:` syntax, roughly `"//example.com/ns/busybox:latest"`); that would allow you to get rid of quite a few dependencies.

Comment (ilackarms) [inline Gopkg.lock]: @enoodle @simon3z thoughts on this? personally my preference is to include more dependencies to allow for possible edge cases and preclude technical debt (i.e. having to come in and change this back to `alltransports` at a future time).

File: pkg/inspector/image-inspector.go (hunk 10 above)
Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: i don't believe it is possible to have a panic here (even in the case of empty string, this function returns `[]string{""}` which won't cause a panic anywhere below

File: Dockerfile (hunk 3 above)
Comment (ilackarms) [inline Dockerfile]: true. this will be resolved anyway once we merge https://github.com/openshift/image-inspector/pull/59 this will be irrelevant; however until then i'm make sure we remove `go` once the binary is built

**Comment by ilackarms** [comment]
@simon3z the deps have already been pruned; unfortunately all the added dependencies appear to be used (at least according to `dep`)

**Comment by ilackarms** [comment]
@pweil- @mfojtik can you please review?

**Comment by pweil-** [comment]
I'm currently slammed with 3.6 closing items.  Will have to get to it after that unless someone else can give it the :+1: earlier

File: pkg/inspector/image-inspector.go [hunk 18]
```diff
@@ -533,16 +480,164 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+		reportBuffer := &bytes.Buffer{}
+		err = copy.Image(policyContext, destRef, srcRef, &copy.Options{
+			RemoveSignatures: false,
+			SignBy:           "",
+			ReportWriter:     reportBuffer,
+			SourceCtx:        sourceCtx,
+			DestinationCtx:   nil,
+			ProgressInterval: PULL_LOG_INTERVAL_SEC,
+		})
+		log.Printf("Output of copy image operation using auth %s:\n%s", name, reportBuffer.String())
```
Comment (enoodle) [inline pkg/inspector/image-inspector.go]: @ilackarms We should use `reportBuffer` to print the ongoing pull progress to the user. See https://github.com/openshift/image-inspector/issues/13 and its solution , also in this file see if you can redirect the output to `reportBuffer` to `aggregateBytesAndReport`  that will run in a parallel go routine.

Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: I don't believe we can use `aggregateBytesAndReport` since `image.Copy` doesn't provide us any structured data about the transfer progress (rather it's just an output writer for logging)

Comment (mtrmac) [inline pkg/inspector/image-inspector.go]: Actually there is `copy.Options.ProgressInterval` and `copy.Options.Progress`. I don’t know whether it is good enough, if it’s not, PRs to improve it would be welcome.

**Comment by ilackarms** [comment]
note: we are waiting for openshift/origin#15298 to enable running pod without any special SCC / privilege

**Comment by ilackarms** [comment]
update: https://github.com/openshift/origin/pull/15298 has been merged. we should be able to run image inspector without any special privileges after this patch is merged @simon3z

**Comment by hypery2k** [comment]
@ilackarms Do you plan to update this PR?

**Comment by ilackarms** [comment]
@hypery2k not sure this project is being actively developed anymore (i'm no longer a part of it). maybe @pweil- knows the current maintainers?

**Comment by pweil-** [comment]
@simon3z can you comment here?

**Comment by simon3z** [comment]
> @simon3z can you comment here?

@pweil- I am available for high level review, design guidance, etc.

For active development of the project you should sync with @oourfali

**Comment by zeari** [comment]
@hypery2k We are not going forward with this PR at this time...

**Comment by hypery2k** [comment]
okay, thanks for the update. Are there plans to support running without root rights?

**Comment by simon3z** [comment]
> okay, thanks for the update. Are there plans to support running without root rights?

@hypery2k there are several half-baked PRs that were aiming at that. All in all I think the code needed to achieve is in the PRs but they need grooming, testing, reviews, etc.

**Comment by hypery2k** [comment]
okay, i could support on the testing. Did you have any PR that looks most promising? I could even test several of them

**Comment by simon3z** [comment]
> okay, i could support on the testing. Did you have any PR that looks most promising? I could even test several of them

@hypery2k IIRC I think #87 and #95 (maybe #95 already contains #87)
